File: sdk/ruby/test/wire/billing_test.rb
# frozen_string_literal: true

require_relative "wiremock_test_case"

class BillingWireTest < WireMockTestCase
  def setup
    super

    @client = AgentTrunk::Client.new(
      access_token: "<token>",
      base_url: WIREMOCK_BASE_URL
    )
  end

  def test_billing_get_with_wiremock
    test_id = "billing.get.0"

    @client.billing.get(request_options: {
      additional_headers: {
        "X-Test-Id" => "billing.get.0"
      }
    })

    verify_request_count(
      test_id: test_id,
      method: "GET",
      url_path: "/v1/billing",
      query_params: nil,
      expected: 1
    )
  end

  def test_billing_create_checkout_with_wiremock
    test_id = "billing.create_checkout.0"

    @client.billing.create_checkout(
      plan: "starter",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "billing.create_checkout.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/v1/billing/checkout",
      query_params: nil,
      expected: 1
    )
  end

  def test_billing_create_portal_with_wiremock
    test_id = "billing.create_portal.0"

    @client.billing.create_portal(request_options: {
      additional_headers: {
        "X-Test-Id" => "billing.create_portal.0"
      }
    })

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/v1/billing/portal",
      query_params: nil,
      expected: 1
    )
  end

  def test_billing_set_spend_limit_with_wiremock
    test_id = "billing.set_spend_limit.0"

    @client.billing.set_spend_limit(
      cents: 1,
      request_options: {
        additional_headers: {
          "X-Test-Id" => "billing.set_spend_limit.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/v1/billing/spend-limit",
      query_params: nil,
      expected: 1
    )
  end
end

File: sdk/rust/tests/billing_test.rs
use agenttrunk::prelude::*;

mod wire_test_utils;

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_billing_get_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentTrunk::new(config).expect("Failed to build client");

    let result = client.billing.get(None).await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("GET", "/v1/billing", None, 1)
        .await
        .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_billing_create_checkout_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentTrunk::new(config).expect("Failed to build client");

    let result = client
        .billing
        .create_checkout(
            &CreateCheckoutBillingRequest {
                plan: CreateCheckoutBillingRequestPlan::Starter,
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("POST", "/v1/billing/checkout", None, 1)
        .await
        .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_billing_create_portal_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentTrunk::new(config).expect("Failed to build client");

    let result = client.billing.create_portal(None).await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("POST", "/v1/billing/portal", None, 1)
        .await
        .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_billing_set_spend_limit_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentTrunk::new(config).expect("Failed to build client");

    let result = client
        .billing
        .set_spend_limit(&SetSpendLimitBillingRequest { cents: 1 }, None)
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("POST", "/v1/billing/spend-limit", None, 1)
        .await
        .unwrap();
}
